File: backend/src/main/java/com/gxs/bank/model/Loan.java
package com.gxs.bank.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "loans")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Loan {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @JsonIgnore
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(name = "loan_type", nullable = false)
    private LoanType loanType;

    @Column(name = "loan_name")
    private String loanName;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal amount;

    @Column(name = "outstanding_amount", precision = 15, scale = 2)
    private BigDecimal outstandingAmount;

    @Column(name = "interest_rate", nullable = false, precision = 5, scale = 2)
    private BigDecimal interestRate;

    @Column(name = "tenure_months", nullable = false)
    private Integer tenureMonths;

    @Column(name = "monthly_payment", precision = 15, scale = 2)
    private BigDecimal monthlyPayment;

    @Column(name = "total_emis")
    private Integer totalEmis;

    @Column(name = "emis_paid")
    private Integer emisPaid;

    @Column(name = "next_emi_date")
    private LocalDate nextEmiDate;

    @Column(name = "processing_fee", precision = 15, scale = 2)
    private BigDecimal processingFee;

    @Column(name = "purpose")
    private String purpose;

    @Column(name = "applied_by")
    private UUID appliedBy;

    @Column(name = "approved_by")
    private UUID approvedBy;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "disbursed_account_id")
    private UUID disbursedAccountId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Status status;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (status == null) status = Status.PENDING;
        if (outstandingAmount == null) outstandingAmount = amount;
        if (totalEmis == null) totalEmis = tenureMonths;
        if (emisPaid == null) emisPaid = 0;
        if (processingFee == null) processingFee = BigDecimal.ZERO;
    }

    public enum LoanType { PERSONAL, HOME, VEHICLE, EDUCATION, GOLD, BUSINESS, BALANCE_TRANSFER, INSTALMENT }
    public enum Status { PENDING, ACTIVE, PAID_OFF, DEFAULTED, REJECTED }
}

File: backend-python/com/gxs/bank/model/Loan.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
import uuid

from sqlalchemy import Column, Date, DateTime, Enum as SQLEnum, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import relationship

from com.gxs.bank.config.database import Base


class LoanType(str, Enum):
    PERSONAL = "PERSONAL"
    HOME = "HOME"
    VEHICLE = "VEHICLE"
    EDUCATION = "EDUCATION"
    GOLD = "GOLD"
    BUSINESS = "BUSINESS"
    BALANCE_TRANSFER = "BALANCE_TRANSFER"
    INSTALMENT = "INSTALMENT"


class Status(str, Enum):
    PENDING = "PENDING"
    ACTIVE = "ACTIVE"
    PAID_OFF = "PAID_OFF"
    DEFAULTED = "DEFAULTED"
    REJECTED = "REJECTED"


class Loan(Base):
    __tablename__ = "loans"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    userId = Column("user_id", String(36), ForeignKey("users.id"), nullable=False)
    user = relationship("User", lazy="select")

    loanType = Column("loan_type", SQLEnum(LoanType), nullable=False)
    loanName = Column("loan_name", String(255), nullable=True)
    amount = Column(Numeric(15, 2), nullable=False)
    outstandingAmount = Column("outstanding_amount", Numeric(15, 2), nullable=True)
    interestRate = Column("interest_rate", Numeric(5, 2), nullable=False)
    tenureMonths = Column("tenure_months", Integer, nullable=False)
    monthlyPayment = Column("monthly_payment", Numeric(15, 2), nullable=True)
    totalEmis = Column("total_emis", Integer, nullable=True)
    emisPaid = Column("emis_paid", Integer, nullable=True)
    nextEmiDate = Column("next_emi_date", Date, nullable=True)
    processingFee = Column("processing_fee", Numeric(15, 2), nullable=False, default=Decimal("0.00"))
    purpose = Column("purpose", String(1024), nullable=True)
    appliedBy = Column("applied_by", String(36), nullable=True)
    approvedBy = Column("approved_by", String(36), nullable=True)
    approvedAt = Column("approved_at", DateTime, nullable=True)
    disbursedAccountId = Column("disbursed_account_id", String(36), nullable=True)
    status = Column(SQLEnum(Status), nullable=False, default=Status.PENDING)
    createdAt = Column("created_at", DateTime, nullable=False, default=datetime.utcnow)


Loan.LoanType = LoanType
Loan.Status = Status
